les/odr_requests.go

```diff
@@ -390,73 +397,88 @@ func (r *ChtRequest) Validate(db ethdb.Database, msg *Msg) error {
 
 	switch msg.MsgType {
 	case MsgHeaderProofs: // LES/1 backwards compatibility
-		proofs := msg.Obj.([]ChtResp)
-		if len(proofs) != 1 {
+		resps := msg.Obj.([]ChtResp)
+		if len(resps) != len(r.BlockNum) {
 			return errInvalidEntryCount
 		}
-		proof := proofs[0]
-
-		// Verify the CHT
-		var encNumber [8]byte
-		binary.BigEndian.PutUint64(encNumber[:], r.BlockNum)
-
-		value, _, err := trie.VerifyProof(r.ChtRoot, encNumber[:], light.NodeList(proof.Proof).NodeSet())
-		if err != nil {
-			return err
-		}
-		var node light.ChtNode
-		if err := rlp.DecodeBytes(value, &node); err != nil {
-			return err
+		var (
+			headers   []*types.Header
+			tds       []*big.Int
+			encNumber [8]byte
+			node      light.ChtNode
+			nodeset   = light.NewNodeSet()
+		)
+		for i, num := range r.BlockNum {
+			resp := resps[i]
+			// Verify the CHT
+			binary.BigEndian.PutUint64(encNumber[:], num)
+			value, _, err := trie.VerifyProof(r.ChtRoot, encNumber[:], light.NodeList(resp.Proof).NodeSet())
+			if err != nil {
+				return err
+			}
+			if err := rlp.DecodeBytes(value, &node); err != nil {
+				return err
+			}
+			if node.Hash != resp.Header.Hash() {
+				return errCHTHashMismatch
+			}
+			if num != resp.Header.Number.Uint64() {
+				return errCHTNumberMismatch
+			}
+			// Verifications passed, store temporarily
+			headers = append(headers, resp.Header)
+			tds = append(tds, node.Td)
+			light.NodeList(resp.Proof).Store(nodeset)
 		}
-		if node.Hash != proof.Header.Hash() {
-			return errCHTHashMismatch
-		}
-		// Verifications passed, store and return
-		r.Header = proof.Header
-		r.Proof = light.NodeList(proof.Proof).NodeSet()
-		r.Td = node.Td
+		r.Header = headers
+		r.Td = tds
+		r.Proof = nodeset
+
 	case MsgHelperTrieProofs:
+		// Check if the number of items in the response is the same as we requested.
 		resp := msg.Obj.(HelperTrieResps)
-		if len(resp.AuxData) != 1 {
+		if len(resp.AuxData) != len(r.BlockNum) {
 			return errInvalidEntryCount
 		}
-		nodeSet := resp.Proofs.NodeSet()
-		headerEnc := resp.AuxData[0]
-		if len(headerEnc) == 0 {
-			return errHeaderUnavailable
-		}
-		header := new(types.Header)
-		if err := rlp.DecodeBytes(headerEnc, header); err != nil {
-			return errHeaderUnavailable
+		var (
+			headers   []*types.Header
+			tds       []*big.Int
+			encNumber [8]byte
+			node      light.ChtNode
+			nodeSet   = resp.Proofs.NodeSet()
+		)
+		for i, num := range r.BlockNum {
+			enc := resp.AuxData[i]
+			if len(enc) == 0 {
+				return errHeaderUnavailable
+			}
+			header := new(types.Header)
+			if err := rlp.DecodeBytes(enc, header); err != nil {
+				return errHeaderUnavailable
+			}
+			// Verify the CHT
+			binary.BigEndian.PutUint64(encNumber[:], num)
+			value, _, err := trie.VerifyProof(r.ChtRoot, encNumber[:], nodeSet)
```

Comment:
In the old version I was checking for useless nodes with &readTraceDB{db: nodeSet} which tracks whether all nodes have been read by the verifier. Please do the same here by creating readTraceDB before the cycle and checking whether len(reads.reads) != nodeSet.KeyCount() once after all block proofs have been verified.

Reply:
I was trying to add the similar logic before i pushed up, but i found some nodes will be read many times(e.g. root node). So maybe the checking will not work here.